src/cli/cli_tcp.cpp

```diff
@@ -0,0 +1,522 @@
+                                                                       otTcpEndpoint **  aAcceptInto)
+{
+    return static_cast<TcpExample *>(otTcpListenerGetContext(aListener))
+        ->HandleTcpAcceptReady(aListener, aPeer, aAcceptInto);
+}
+
+void TcpExample::HandleTcpAcceptDoneCallback(otTcpListener *   aListener,
+                                             otTcpEndpoint *   aEndpoint,
+                                             const otSockAddr *aPeer)
+{
+    static_cast<TcpExample *>(otTcpListenerGetContext(aListener))->HandleTcpAcceptDone(aListener, aEndpoint, aPeer);
+}
+
+void TcpExample::HandleTcpEstablished(otTcpEndpoint *aEndpoint)
+{
+    OT_UNUSED_VARIABLE(aEndpoint);
+    mInterpreter.OutputLine("TCP: Connection established");
+}
+
+void TcpExample::HandleTcpSendDone(otTcpEndpoint *aEndpoint, otLinkedBuffer *aData)
+{
+    OT_UNUSED_VARIABLE(aEndpoint);
+
+    if (mBenchmarkBytesTotal == 0)
+    {
+        /*
+         * If the benchmark encountered an error, we might end up here. So,
+         * tolerate some benchmark links finishing in this case.
+         */
+        if (aData == &mSendLink)
+        {
+            OT_ASSERT(mSendBusy);
+            mSendBusy = false;
+        }
+    }
+    else
+    {
+        OT_ASSERT(aData != &mSendLink);
+        mBenchmarkLinksLeft--;
+        if (mBenchmarkLinksLeft >= OT_ARRAY_LENGTH(mBenchmarkLinks))
+        {
+            if (otTcpSendByReference(&mEndpoint, aData, 0) != OT_ERROR_NONE)
+            {
+                mInterpreter.OutputLine("TCP Benchmark Failed");
+                mBenchmarkBytesTotal = 0;
+            }
+        }
+        else if (mBenchmarkLinksLeft == 0)
+        {
+            uint32_t milliseconds         = TimerMilli::GetNow() - mBenchmarkStart;
+            uint32_t thousandTimesGoodput = (1000 * (mBenchmarkBytesTotal << 3) + (milliseconds >> 1)) / milliseconds;
+            mInterpreter.OutputLine("TCP Benchmark Complete: Transferred %u bytes in %u milliseconds",
+                                    static_cast<unsigned int>(mBenchmarkBytesTotal),
+                                    static_cast<unsigned int>(milliseconds));
+            mInterpreter.OutputLine("TCP Goodput: %u.%03u kb/s", thousandTimesGoodput / 1000,
+                                    thousandTimesGoodput % 1000);
+            mBenchmarkBytesTotal = 0;
+        }
+    }
+}
+
+void TcpExample::HandleTcpReceiveAvailable(otTcpEndpoint *aEndpoint,
+                                           size_t         aBytesAvailable,
+                                           bool           aEndOfStream,
+                                           size_t         aBytesRemaining)
+{
+    OT_UNUSED_VARIABLE(aBytesRemaining);
+    OT_ASSERT(aEndpoint == &mEndpoint);
+
+    if (aBytesAvailable > 0)
+    {
+        const otLinkedBuffer *data;
+        IgnoreError(otTcpReceiveByReference(aEndpoint, &data));
+        size_t totalReceived = 0;
```

Comment:
All variable should be toghther at the top of scope.

```suggestion
        const otLinkedBuffer *data;
        size_t totalReceived = 0;

        IgnoreError(otTcpReceiveByReference(aEndpoint, &data));
```